### ISSUE TITLE
Builder freezes with dart.exe running at 100% CPU

### ISSUE THREAD

**Issue by JCKodel** [simolus3/moor#2389]
**Describe the bug**

I am tinkering with a way to create a remote database sync using triggers. These particular drift files makes build_runner freeze with dart consuming 100% cpu:

<details><summary>sync.drift</summary>

```sql
CREATE TABLE "SyncChanges" (
  "table"           TEXT      NOT NULL,
  "pk"              TEXT      NOT NULL,
  "lastUpdateTime"  DATETIME  NOT NULL,
  "isDeleted"       BOOLEAN   NOT NULL,

  PRIMARY KEY("table", "pk")
);
```

</details>

<details><summary>users.drift</summary>

```sql
import 'sync.drift';
import 'users.drift';
import 'auth_provider.dart';
import 'package:kfx/kfx/environment.dart';

CREATE TABLE "Users" (
  "id"            TEXT      NOT NULL,
  "isBlocked"     BOOLEAN   NOT NULL,
  "displayName"   TEXT      NOT NULL,
  "avatarURL"     TEXT      NOT NULL,
  "email"         TEXT      NOT NULL,
  "creationTime"  DATETIME  NOT NULL,

  PRIMARY KEY("id")
);

CREATE TABLE UserLogins (
  "userId"          TEXT                      NOT NULL,
  "deviceId"        TEXT                      NOT NULL,
  "authProvider"    ENUMNAME(AuthProvider)    NOT NULL,
  "loginDate"       DATETIME                  NOT NULL,
  "model"           TEXT                      NOT NULL,
  "device"          TEXT                      NOT NULL,
  "version"         TEXT                      NOT NULL,
  "nativePlatform"  ENUMNAME(NativePlatform)  NOT NULL,

  PRIMARY KEY("userId", "deviceId"),
  FOREIGN KEY("userId") REFERENCES Users("id")
);

CREATE INDEX "UserLogins_loginDate" ON "UserLogins"("loginDate" DESC);

CREATE TRIGGER "Users_insert" AFTER INSERT ON "Users" BEGIN
  INSERT INTO 
    "SyncChanges"("table", "pk", "lastUpdateTime", "isDeleted")
    VALUES ('Users', '{"id": "' || old."id" || '"}', strftime('%Y-%m-%dT%H:%M:%fZ', 'now'), 0)
  ON CONFLICT("table", "pk") DO UPDATE SET
    "lastUpdateTime" = excluded."lastUpdateTime",
    "isDeleted" = excluded."isDeleted";
END;

CREATE TRIGGER "Users_update" AFTER UPDATE ON "Users" 
WHEN
  old."isBlocked" <> new."isBlocked"
  OR old."displayName" <> new."displayName"
  OR old."avatarURL" <> new."avatarURL"
  OR old."email" <> new."email"
BEGIN
  INSERT INTO 
    "SyncChanges"("table", "pk", "lastUpdateTime", "isDeleted")
    VALUES ('Users', '{"id": "' || old."id" || '"}', strftime('%Y-%m-%dT%H:%M:%fZ', 'now'), 0)
  ON CONFLICT("table", "pk") DO UPDATE SET
    "lastUpdateTime" = excluded."lastUpdateTime",
    "isDeleted" = excluded."isDeleted";
END;

CREATE TRIGGER "Users_delete" AFTER DELETE ON "Users" BEGIN
  INSERT INTO 
    "SyncChanges"("table", "pk", "lastUpdateTime", "isDeleted")
    VALUES ('Users', '{"id": "' || old."id" || '"}', strftime('%Y-%m-%dT%H:%M:%fZ', 'now'), 1)
  ON CONFLICT("table", "pk") DO UPDATE SET
    "lastUpdateTime" = excluded."lastUpdateTime",
    "isDeleted" = excluded."isDeleted";
END;

CREATE TRIGGER "UserLogins_insert" AFTER INSERT ON "UserLogins" BEGIN
  INSERT INTO 
    "SyncChanges"("table", "pk", "lastUpdateTime", "isDeleted")
    VALUES ('UserLogins', '{"userId": "' || new."userId" || '", "deviceId": "' || new."deviceId" || '"}', strftime('%Y-%m-%dT%H:%M:%fZ', 'now'), 0);
  ON CONFLICT("table", "pk") DO UPDATE SET
    "lastUpdateTime" = excluded."lastUpdateTime",
    "isDeleted" = excluded."isDeleted";
END;

CREATE TRIGGER "UserLogins_update" AFTER UPDATE ON "UserLogins" 
WHEN 
  old."authProvider" <> new."authProvider"
  OR strftime('%Y-%m-%dT%H:%M:%fZ', old."loginDate") <> strftime('%Y-%m-%dT%H:%M:%fZ', new."loginDate")
  OR old."model" <> new."model"
  OR old."device" <> new."device"
  OR old."version" <> new."version"
  OR old."nativePlatform" <> new."nativePlatform"
BEGIN
  INSERT INTO 
    "SyncChanges"("table", "pk", "lastUpdateTime", "isDeleted")
    VALUES ('UserLogins', '{"userId": "' || new."userId" || '", "deviceId": "' || new."deviceId" || '"}', strftime('%Y-%m-%dT%H:%M:%fZ', 'now'), 0);
  ON CONFLICT("table", "pk") DO UPDATE SET
    "lastUpdateTime" = excluded."lastUpdateTime",
    "isDeleted" = excluded."isDeleted";
END;

CREATE TRIGGER "UserLogins_delete" AFTER DELETE ON "UserLogins" BEGIN
  INSERT INTO 
    "SyncChanges"("table", "pk", "lastUpdateTime", "isDeleted")
    VALUES ('UserLogins', '{"userId": "' || new."userId" || '", "deviceId": "' || new."deviceId" || '"}', strftime('%Y-%m-%dT%H:%M:%fZ', 'now'), 1);
  ON CONFLICT("table", "pk") DO UPDATE SET
    "lastUpdateTime" = excluded."lastUpdateTime",
    "isDeleted" = excluded."isDeleted";
END;

getUserById:
  SELECT * FROM "Users" WHERE "id" = :id LIMIT 1;

getLastUserSignIn:
  SELECT * FROM "UserLogins" WHERE "userId" = :userId
  ORDER BY "loginDate" DESC
  LIMIT 1;

saveUser:
  INSERT INTO 
    "Users"("id", "isBlocked", "displayName", "avatarURL", "email", "creationTime") 
    VALUES(:id, :isBlocked, :displayName, :avatarURL, :email, :creationTime)
  ON CONFLICT("id") DO UPDATE SET
    "isBlocked" = excluded."isBlocked",
    "displayName" = excluded."displayName",
    "avatarURL" = excluded."avatarURL",
    "email" = excluded."email";

saveUserLogin:
  INSERT INTO 
    "UserLogins"("userId", "authProvider", "loginDate", "deviceId", "model", "device", "version", "nativePlatform")
    VALUES(:userId, :authProvider, :loginDate, :deviceId, :model, :device, :version, :nativePlatform)
  ON CONFLICT("userId", "deviceId") DO UPDATE SET
    "authProvider" = excluded."authProvider",
    "loginDate" = excluded."loginDate",
    "deviceId" = excluded."deviceId",
    "model" = excluded."model",
    "device" = excluded."device",
    "version" = excluded."version",
    "nativePlatform" = excluded."nativePlatform";
```

</details>

The package kfx is an external package that uses `part of` to join some dart files (in which one of them contains an Enum that I'm using in the drift file):

<details><summary>environment.dart</summary>

```dart
library kfx_environment;

import 'package:flutter/widgets.dart';

import 'environment/web_platform_info.dart' if (dart.library.io) "environment/io_platform_info.dart";

part "environment/environment.dart";
part "environment/native_platform.dart";
part "environment/platform_design_system.dart";
part "environment/platform_host.dart";
part "environment/platform_info.dart";
part "environment/platform_media.dart";
```

</details>

I'm using modular build, and this is my build.yaml:

<details><summary>build.yaml</summary>

```yaml
targets:
  $default:
    builders:
      drift_dev:
        enabled: false

      drift_dev:analyzer:
        enabled: true
        options: &options
          store_date_time_values_as_text: true
          override_hash_and_equals_in_result_sets: true
          named_parameters: true

      drift_dev:modular:
        enabled: true
        options: *options
```

</details>

My database file:

<details><summary>database.dart</summary>

```dart
import 'package:drift/drift.dart';

import 'database.drift.dart';
import 'web_database.dart' if (dart.library.io) "io_database.dart";

@DriftDatabase(
  include: {
    "sync.drift",
    "users.drift",
  },
)
class Database extends $Database {
  Database._() : super(openConnection());

  @override
  int get schemaVersion => 1;

  static final Database _instance = Database._();
  static Database get instance => _instance;
}
```

</details>

BTW: removing all triggers "fixes" the issue, so there is something wrong with the parser in that point.

**Comment by simolus3** [comment]
Thanks for the report! The problem is a syntax error in the `UserLogins_*` triggers: You have a semicolon after the `VALUES` clause but the statement doesn't end there (there should only be a semicolon after the on conflict clause).

```sql
CREATE TRIGGER "UserLogins_delete" AFTER DELETE ON "UserLogins" BEGIN
  INSERT INTO 
    "SyncChanges"("table", "pk", "lastUpdateTime", "isDeleted")
    VALUES ('UserLogins', '{"userId": "' || new."userId" || '", "deviceId": "' || new."deviceId" || '"}', strftime('%Y-%m-%dT%H:%M:%fZ', 'now'), 1);
                                                                                                                       this needs to be removed!   |
  ON CONFLICT("table", "pk") DO UPDATE SET
    "lastUpdateTime" = excluded."lastUpdateTime",
    "isDeleted" = excluded."isDeleted";
END;
```

Of course, that shouldn't cause the parser to hit an infinite loop.

**Comment by simolus3** [comment]
Fixed in 99abaae2f1449ef8feebdf00b27a3fe91b8a1cd1!
